Add gateway architecture page to docs (#3008)
hyperledger/fabric: docs/source/gateway.md

> # Gateway

Reviewer: "Fabric Gateway"

Reviewer: done.

> 

Reviewer: (Generally, docs seek to avoid "empty" headings- i.e. ideally H1s should also have some text.) So suggest an add of something like: 

"Fabric v2.4 introduces the new Fabric Gateway, which implements functions previously required of client applications, resulting in a simplified API and consumable SDKs. Functions migrated from SDKs to the new Gateway are ... (list here), resulting in (add a few process and/or business benefits)."

(Any benefits to security that we can promote/list? or other benefits.)

Reviewer: done.

## Overview

The Gateway is a new service introduced in v2.4 that provides a simplified, minimal API for submitting transactions
to a Fabric network.

## Writing client applications

> Starting with v2.4, client applications should use one of the v2.4 SDKs which are optimised to interact with the Gateway.

Reviewer: "... one of the v2.4 SDKs (Go, Java, or Node), which are optimized to ..."

Reviewer: done.

These SDKs expose the same high-level programming model which was first introduced with v1.4 of Fabric.

> The gateway supports the following actions

Reviewer: "The Gateway supports the following actions:" 
(blank line) 
(Here I was expecting a "Proposal" transaction step but perhaps that is still up to the client app right. idk if worth mentioning proposal.)

Reviewer: Also it "supports" them or it actually "does/completes/runs" them. 
If sequential order, use numbers instead of bullets.

Reviewer: "manages the following transaction steps:" 
no sequential order

> - **Evaluate** (query) a transaction.  This will invoke a smart contract (chaincode) function on a single peer and return the result

Reviewer 2: "Query a transaction" doesn't sound right... maybe:

```suggestion
- **Evaluate** (query) a chaincode.  This will invoke a smart contract (chaincode) function on a single peer and return the result
```

Reviewer: No Tx proposal here right, that's before entering the Gateway?

Author: True.  I've changed it to:
   - **Evaluate** a transaction proposal.  This will invoke a smart contract (chaincode) function on a single peer and return the result to the client.  This is typically used to query the current state of the ledger without making any ledger updates.

> to the client.  The gateway will preferably select a peer in the same organization as the gateway peer and choose the peer 

Reviewer: "The Gateway will first attempt to select a peer in the organization that owns the gateway peer. If no peer is available, it will select a peer from another organization. In either case, the Gateway will select the peer with the highest block height (on its ledger)."

Reviewer: i.e. "preferably" can be read like we hope it does.

with the highest ledger block height.  If no peer is available in the gateway's organization, then it will choose a peer
from another organization.

> - **Endorse** a transaction. This will gather enough endorsements to satisfy the combined signature policies (see below) and

Reviewer 2: ```suggestion
- **Endorse** a transaction proposal. This will gather enough endorsements to satisfy the combined signature policies (see below) and
```

Reviewer: It "will" gather or "attempt" to gather? 
idk if gather is the best word. collect, retrieve ...

Reviewer: Is it always "enough" endorsements that is required, or is it sometimes also an endorsement from a certain trusted node?

> return a prepared transaction envelope to the client for signing.

Reviewer 2: I'm thinking we should include the word "response" here since we talk about proposal response in other parts of the doc. However, I'm forgetting exactly how EndorserResponse, ProposalResponse, and PreparedTransaction are related. Could you remind me :-)

Author: Changed to `This will gather enough endorsement responses to satisfy...`
The various response structures you mention are just the protobuf messages that get returned from the various gateway grpc methods.  They are wrappers for the peer ProposalResponse, etc messages and contain some other information. Details here https://github.com/hyperledger/fabric-protos/blob/main/gateway/gateway.proto
I wasn't intending to document the proto level API in this docs topic.

Reviewer: ok.

> - **Submit** a transaction.  This will send a signed transaction envelope to the ordering service to commit to the ledger.

Reviewer: Okay so client has to do something to sign and return the Tx envelope right. But then we're saying the Gateway does the "submit"?

Reviewer: ignore.

- Wait for **commit status** events.  This allows the client to wait for the transaction to be committed to the ledger and to

> get the commit (success/failure) status code.

Reviewer 2: ```suggestion
get the commit (validation/invalidation) status code.
```

- Receive **chaincode events**. This will allow client applications to respond to events that are emitted by a smart contract

> function when a transaction is committed to the ledger.

Reviewer: only on commit step?

Reviewer: ignore.

> The gateway SDKs will combine the Endorse/Submit/CommitStatus actions into a single blocking SubmitTransaction function to support transaction

Reviewer 2: ```suggestion
The gateway SDKs combine the Endorse/Submit/CommitStatus actions into a single blocking SubmitTransaction function to support transaction
```

submission with a single line of code.  Alternatively, the individual actions can be invoked to support flexible application patterns.

## Software Development Kits (SDKs)

Hyperledger Fabric currently supports client application development in three languages:

- **Go**.  See the [Go SDK documentation](https://pkg.go.dev/github.com/hyperledger/fabric-gateway/pkg/client) for full details.
- **Node (Typescript/Javascript)**.  See the [Node SDK documentation](https://hyperledger.github.io/fabric-gateway/main/api/node/) for full details.
- **Java**. See the [Java SDK documentation](https://hyperledger.github.io/fabric-gateway/main/api/java/) for full details.

## How the gateway endorses your transaction proposal

> In order for a transaction to be successfully committed to the ledger, a sufficient number of endorsements are required in order to satisfy

Reviewer: it's only number of ... ? ... it doesn't require the gateway peer, e.g. or other trusted peer, or a different endorsement for SBEs?

Reviewer: ignore, number implies orgs here.

the [endorsement policy](endorsement-policies.html).  Getting an endorsement from an organization involves connecting to one 

> of its peers and have it simulate (execute) the transaction proposal against its copy of the ledger.  The peer does this by

Reviewer: ```suggestion
of its peers and have it simulate (execute) the transaction proposal against its copy of the ledger.  The peer simulates the transaction by
```

Reviewer: done.

> invoking the chaincode function whose name and arguments are specified in the proposal and building (and signing) a read-write set

Reviewer: ```suggestion
invoking the chaincode function, as specified by its name and arguments in the proposal, and building (and signing) a read-write set. 
```

Reviewer: done.

> containing the current ledger state and changes it would make in response to the state get/set instructions in that function.

Reviewer: ```suggestion
The read-write set contains the current ledger state and proposed changes in response to the state get/set instructions in that function.
```

Reviewer: done.

> The endorsement policy that gets applied to a transaction depends on the implementation of the chaincode function that 

Reviewer: ```suggestion
The endorsement policy, or sum of multiple policies, that gets applied to a transaction depends on the implementation of the chaincode function that 
```

Reviewer: done.

is being invoked, and could be a combination of the following:

> - **Chaincode policies**.  These are the policies agreed to by channel members when they approve a chaincode definition for their organization.

Reviewer: ```suggestion
- **Chaincode endorsement policies**.  These are the policies agreed to by channel members when they approve a chaincode definition for their organization.
```

Reviewer: done.

If the chaincode function invokes a function in another chaincode, then both policies will need to be satisfied.

> - **Private data collection policies**.  If the chaincode function writes to a state in a private data collection,

Reviewer: ```suggestion
- **Private data collection endorsement policies**.  If the chaincode function writes to a state in a private data collection,
```

Reviewer: done.

> then the endorsement policy for that collection will override the chaincode policy for that state.  If it reads from a private

Reviewer: ```suggestion
then the endorsement policy for that collection will override the chaincode policy for that state.  If the chaincode function reads from a private data 
```

Reviewer: done.

> collection, then it will be restricted to organizations that own a copy of that collection.

Reviewer: In line 52, do we really mean "peers" here rather than the "organizations" that own them?

- **State-based endorsement (SBE) policies**. Also known as key-level signature policies, these can be applied to individual

> states and will override the chaincode policy (or collection policy if it's a private collection state).

Reviewer: ```suggestion
states and will override the chaincode policy, or collection policy for private data  collection states.
```

Reviewer: done.

> The policies themselves are stored in the ledger and can be modified by transactions.

Reviewer: ```suggestion
The endorsement policies themselves are stored in the ledger and can be modified by transactions.
```

Reviewer: done.

Although it might seem obvious from the transaction proposal which policies would be applied, the actual combination of
polices depends on what the chaincode function does at runtime and cannot necessarily be derived from static analysis.

The gateway handles all of this complexity on behalf of the client using the following process:

- The gateway selects the endorsing peer in the gateway's organization (MSPID) that has the highest ledger block height.
The assumption is that all peers within this organization are *trusted* by the client application that connects
to this gateway.
- It simulates the transaction proposal on that peer.  During simulation, it captures all the relevant information on which
states are accessed and hence which policies will need to be combined (including any individual state-based policies 
as stored in the peer's copy of the ledger).  
- This information is assembled into a `ChaincodeInterest` protobuf structure and passed to the discovery service in order

> to derive an endorsement plan specific for this transaction.

Reviewer 2: I think we should mention the dependency on discovery service prior to this point. Provide a link to discovery service and mention this is how gateway becomes aware of available ordering service nodes and peers.

Reviewer: ```suggestion
to derive an endorsement plan specific to the proposed transaction.
```

Reviewer: done.

- Using the endorsement plan, the gateway will connect to and request endorsement from other organizations' peers to satisfy
the overall endorsement policy.  It will generally request endorsement from the peer with the highest block height for any
given organization.

If the transaction proposal contains transient data, then the gateway takes a more cautious approach.
Transient data is often used to carry sensitive or personal information that must not persist in the ledger.
A typical usage pattern involves sending the proposal to specific organizations which write the data to their private collections.

> The gateway will restrict the set of endorsing organizations to those that own the collections involved.

Reviewer 2: ```suggestion
The gateway will restrict the set of endorsing organizations to those that are members of the collections involved.
```

Author: I've also changed the wording a bit further up as well to match this.

If this restriction does not satisfy the endorsement policy, it will return an error to the client in preference
to potentially leaking sensitive data.  In this situation, the client should [explicitly define which organizations should 
endorse](#how-to-override-endorsement-policy-detection) the transaction.

### How to override endorsement policy detection

There may be situations where the client application needs to explicitly define which organizations should be targeted to
evaluate or endorse a transaction proposal.  One such example is when sensitive data is being passed in the transient field,
and the client wants to be *absolutely sure* which organizations will receive it.  In these situations the client application
can specify the endorsing organizations. The gateway will select a peer from each of these (bypassing the mechanism described above)
and build a transaction using these endorsements.  It's important to note that if this set of organizations does not satisfy

> the endorsement policy, then validation will fail later after ordering, and a failed transaction will be recorded on the ledger.

Reviewer 2: ```suggestion
the endorsement policy, then the transaction will be included in a block but invalidated by all peers, and the transaction's writes will not be updated in the state database.
```

### Error and retry handling

#### Retry
The gateway, using information from the discovery service, will make every effort to evaluate or submit a transaction in a network
that could suffer from peers or ordering nodes being unavailable for any reason.  If a peer fails, and its organization is 
running multiple peers, then another one will be selected.  If an organization fails to endorse entirely, then another
set of organizations which also satisfies the endorsement policy will be targeted.  Only if there is no combination of available peers
that satisfy the endorsement policy will the gateway stop retrying.  The gateway will not retry the same peer for 

> any given transaction proposal.

Reviewer 2: I have several questions related to this, they don't all necessarily need to be documented, but I think some of them should be.

The gateway will really retry forever if there are many org-peer combinations that satisfy the endorsement policy?
Does the gateway keep the successful endorsements and just try to replace the failed endorsements at the same org(s)?
Or will it gather a whole new set of endorsements potentially using different organizations?
What if you get a chaincode success from one org and a chaincode failure from another org (rather than the peer being unavailable)?

Author: > The gateway will really retry forever if there are many org-peer combinations that satisfy the endorsement policy?

It will try all combinations, but...

> Does the gateway keep the successful endorsements and just try to replace the failed endorsements at the same org(s)?

Yes

> Or will it gather a whole new set of endorsements potentially using different organizations?

No

> What if you get a chaincode success from one org and a chaincode failure from another org (rather than the peer being unavailable)?

It will keep the success, but retry failures, even if it means trying other layouts

Often a 'group' (org) will appear in multiple layouts, so when a successful endorsement is obtained for that group, it could partially satisfy multiple layouts at once.  For every successful endorsement received, it will check the status of all layouts in a plan - it only needs one of them to be satisfied, then it uses that.

Author: Not sure how much of that to document.  This whole section is "good to know", but not essential knowledge for using the gateway.  While it's good to demystify the 'magic', I don't want to tie us into a particular implementation that we can change in the future.

Reviewer: delete, included above.

Reviewer: You can review the updated section in next PR.

# Gateway

> ## Overview

Reviewer: Delete lines 3-6. Included in text above.

Reviewer: done.

is being invoked, and could be a combination of the following:

- **Chaincode policies**.  These are the policies agreed to by channel members when they approve a chaincode definition for their organization.
If the chaincode function invokes a function in another chaincode, then both policies will need to be satisfied.
- **Private data collection policies**.  If the chaincode function writes to a state in a private data collection,
then the endorsement policy for that collection will override the chaincode policy for that state.  If it reads from a private
collection, then it will be restricted to organizations that are members of that collection.
- **State-based endorsement (SBE) policies**. Also known as key-level signature policies, these can be applied to individual
states and will override the chaincode policy (or collection policy if it's a private collection state).
The policies themselves are stored in the ledger and can be modified by transactions.

Although it might seem obvious from the transaction proposal which policies would be applied, the actual combination of

> polices depends on what the chaincode function does at runtime and cannot necessarily be derived from static analysis.

Reviewer: ```suggestion
The combination of endorsement policies to be applied to the transaction proposal is determined at chaincode runtime and cannot necessarily be derived from static analysis.
```

Reviewer: done.

> The gateway handles all of this complexity on behalf of the client using the following process:

Reviewer: ```suggestion
The Fabric Gateway manages the complexity of transaction endorsement on behalf of the client, using the following process:
```

Reviewer: done.

> - The gateway selects the endorsing peer in the gateway's organization (MSPID) that has the highest ledger block height.

Reviewer: If these are sequential use numbered list rather than bullets.

Reviewer: ```suggestion
- The Fabric Gateway selects the endorsing peer from the gateway peer's organization (MSPID) by identifying the (available) peer with the highest ledger block height.
```

Reviewer: done.

> The assumption is that all peers within this organization are *trusted* by the client application that connects

Reviewer: ```suggestion
The assumption is that all peers within the gateway peer's organization are *trusted* by the client application that connects
```

Reviewer: done.

> to this gateway.

Reviewer: ```suggestion
to the gateway peer.
```

Reviewer: done.

> - It simulates the transaction proposal on that peer.  During simulation, it captures all the relevant information on which

Reviewer: ```suggestion
- The transaction proposal is simulated on the selected endorsement peer. This simulation captures information about the accessed states, and therefore the endorsement policies to be combined (including any individual state-based policies 
stored on the endorsement peer's ledger).
```

Reviewer: done.

states are accessed and hence which policies will need to be combined (including any individual state-based policies 
as stored in the peer's copy of the ledger).  

> - This information is assembled into a `ChaincodeInterest` protobuf structure and passed to the discovery service in order

Reviewer: ```suggestion
- The captured policy information is assembled into a `ChaincodeInterest` protobuf structure and passed to the discovery service in order
```

Reviewer: done.

to derive an endorsement plan specific for this transaction.

> - Using the endorsement plan, the gateway will connect to and request endorsement from other organizations' peers to satisfy

Reviewer: ```suggestion
- The gateway applies the endorsement plan by requesting endorsement from the organizations required to satisfy all policies in the plan. For each organization, the gateway peer requests endorsement from the (available) peer with the highest block height.
```

Reviewer: - is "connect to" necessary? or implied. 
- why "generally" what does that mean.

Reviewer: done.

the overall endorsement policy.  It will generally request endorsement from the peer with the highest block height for any
given organization. 
  

> The gateway is dependent on the [discovery service](discovery-overview.html) to get the connection details of available peers

Reviewer: ```suggestion
The gateway is dependent on the [discovery service](discovery-overview.html) to get the connection details of both available peers
```

Reviewer: " ... both the available peers ..." done.

> and ordering service nodes in a network, as well as for calculating the various permutations of peers that are required for 

Reviewer: ```suggestion
and ordering service nodes, and for calculating the combination of peers that are required to endorse the transaction proposal.
```

Reviewer: done.

> endorsement.  It is important, therefore, that the discovery service has not been disabled in a gateway peer.

Reviewer: ```suggestion
The discovery service must therefore always remain enabled on the gateway peer. 
```

Reviewer: "... Fabric Gateway peer." ... done.

> If the transaction proposal contains transient data, then the gateway takes a more cautious approach.

Reviewer: delete line 86, combine with next.

Reviewer: done.

> Transient data is often used to carry sensitive or personal information that must not persist in the ledger.

Reviewer: ```suggestion
The gateway endorsement process is more restrictive for transient data because it often contains sensitive or personal information that must not persist in the ledger.
```

Reviewer: done.

> A typical usage pattern involves sending the proposal to specific organizations which write the data to their private collections.

Reviewer: ```suggestion
The gateway sends proposals involving transient data to only the organizations which write the specific transient data to their private collections.
```

Reviewer: done.

> The gateway will restrict the set of endorsing organizations to those that are members of the collections involved.

Reviewer: ```suggestion
The gateway restricts the endorsing organizations to only those that write the specific transient data to a shared private data collection.
```

Reviewer: Is the concept of "member of a collection" a necessity to externalize? It may be clearer to just state what that means - that this specific transient data (not all transient data) is shared by multiple orgs. ... maybe "share (a copy of) a private data collection" instead of "member of".

Reviewer: done.

> If this restriction does not satisfy the endorsement policy, it will return an error to the client in preference

Reviewer: ```suggestion
If the restriction for transient data results in failed endorsement, the gateway rejects the transaction proposal.
```

Reviewer: done.

to potentially leaking sensitive data.  In this situation, the client should [explicitly define which organizations should 

> endorse](#how-to-override-endorsement-policy-detection) the transaction.

Reviewer: ```suggestion
endorse](#how-to-override-endorsement-policy-detection) the transaction, which must be limited to organizations that share the relevant private data collection.
```

Reviewer: done.

> ### How to override endorsement policy detection

Reviewer: ```suggestion
### Targeting specific endorsement peers
```

Reviewer: Done.

> There may be situations where the client application needs to explicitly define which organizations should be targeted to

Reviewer: ```suggestion
In some cases, a client application must explicitly identify the specific organizations to
```

Reviewer: Done.

> evaluate or endorse a transaction proposal.  One such example is when sensitive data is being passed in the transient field,

Reviewer: ```suggestion
evaluate or endorse a transaction proposal. Transient data, for example, often contains personal or sensitive information that must only be written to a shared private data collection. 
```

Reviewer: Done.

> and the client wants to be *absolutely sure* which organizations will receive it.  In these situations the client application

Reviewer: In these cases, the client application

Reviewer: Done.

> can specify the endorsing organizations. The gateway will select a peer from each of these (bypassing the mechanism described above)

Reviewer: ```suggestion
can explicitly specify the endorsing organizations; the gateway will select the available peer with the highest block count from each organization. 
```

Reviewer: done.

> and build a transaction using these endorsements.  It's important to note that if this set of organizations does not satisfy

Reviewer: ```suggestion
However, if the inclusion of an organization violates a required endorsement policy, then the transaction is invalidated by all of the requested endorsing peers. This failed endorsement is recorded on the ledger but no transaction writes will exist in the state database. 
```

Reviewer: Done.

> the endorsement policy, then the transaction will be included in a block but invalidated by all peers, and the 

Reviewer: included in prior line 100

Reviewer: Done.

> transaction's writes will not be updated in the state database.

Reviewer: included in prior line 100.

Reviewer: Done.

> ### Error and retry handling

Reviewer: ```suggestion
### Retry and error handling
```

Reviewer: Done.

> 

Reviewer: (Some intro words on retry and error.)

Reviewer: Done. "In the case of node unavailability, Fabric Gateway attempts to resolve the transaction request by selecting an available qualifying node."

#### Retry

> The gateway, using information from the discovery service, will make every effort to evaluate or submit a transaction in a network

Reviewer: ```suggestion
The gateway will use discovery service information to retry any transaction that fails due to an unavailable peer or ordering node.
```

Reviewer: Done.

> that could suffer from peers or ordering nodes being unavailable for any reason.  If a peer fails, and its organization is 

Reviewer: ```suggestion
If an organization is running multiple peer or ordering nodes, then another qualifying node will be attempted.
```

Reviewer: done.

> running multiple peers, then another one will be selected.  If an organization fails to endorse entirely, then another

Reviewer: ```suggestion
If an organization fails to endorse a transaction proposal, then another
```

Reviewer: done.

> set of organizations which also satisfies the endorsement policy will be targeted.  Only if there is no combination of available peers

Reviewer: ```suggestion
group of organizations that satisfies the endorsement policy will be targeted.
```

Reviewer: done.

> that satisfy the endorsement policy will the gateway stop retrying.  The gateway will not retry the same peer for 

Reviewer: ```suggestion
The gateway will continue with retry attempts until all possible combinations of endorsing peers have been tried once.
```

Reviewer: done.

any given transaction proposal.

#### Errors

> The gateway manages gPRC connections to other peers and ordering nodes in the network.  If the failure of a gateway service

Reviewer: ```suggestion
The Fabric Gateway manages gRPC connections to peers and ordering nodes in the network. If a gateway service
```

Reviewer: done.

> request is due to an error returned from one or more of these external nodes, then the details of this error, together with

Reviewer: ```suggestion
request error originates from a peer or ordering node (i.e. external to the gateway), the gateway returns error details, endpoint, and organization (MSPID) to the client in the message `Details` field.
```

Reviewer: done.

endpoint address and MSPID of the failing node will be included in the `Details` field of the error returned from the gateway

> to the client.  If the `Details` field is empty, then the error would have originated in the gateway.

Reviewer: ```suggestion
If the `Details` field is empty, then the error originated in the gateway.
```

Reviewer: done.

#### Timeouts

> The gateway `Evaluate` and `Endorse` methods involve making gRPC requests to external peers.  In order to limit the length 

Reviewer: ```suggestion
The Fabric Gateway `Evaluate` and `Endorse` methods make gRPC requests to peers external to the gateway. In order to limit the length 
```

Reviewer: done.

> of time that the client must wait for these collective responses, a configurable `EndorsementTimeout` is available

Reviewer: ```suggestion
of time that the client must wait for these collective responses, the `EndorsementTimeout` value can be overridden 
```

Reviewer: done.

> to override in the gateway section of the [peer config file](https://github.com/hyperledger/fabric/blob/main/sampleconfig/core.yaml#L55).

Reviewer: ```suggestion
in the gateway section of the [peer config file](https://github.com/hyperledger/fabric/blob/main/sampleconfig/core.yaml#L55).
```

Reviewer: done.

> Each SDK also provides a mechanism for setting timeouts for each of the gateway methods when invoked from the client application.

Reviewer: ```suggestion
Each SDK also provides a mechanism for setting timeouts for each gateway method when invoked from the client application.
```

Reviewer: done.

> ## How to listen for events

Reviewer: ```suggestion
## Listening for events
```

Reviewer: done.

> The gateway provides a simplified API for receiving [chaincode events](peer_event_services.html#how-to-register-for-events) 

Reviewer: ```suggestion
The fabric gateway provides a simplified API for client applications to receive [chaincode events](peer_event_services.html#how-to-register-for-events).
```

Reviewer: done.

> in the client applications.  Each SDK provides a mechanism to handle these events using its language-specific idiom.

Reviewer: ```suggestion
Each SDK provides a mechanism to handle these events using its language-specific idiom.
```

Reviewer: done.
